scene/main/node.cpp

```diff
@@ -1346,6 +1346,34 @@ void Node::remove_child(Node *p_child) {
 	remove_child_notify(p_child);
 	p_child->notification(NOTIFICATION_UNPARENTED);
 
+	// Note that calling _set_tree() could have caused
+	// another node to be deleted from the same parent, and thus
+	// idx by this point is not guaranteed to be correct.
+	// We therefore recalculate idx here, for safety, if it has changed.
+	if ((p_child->data.pos != idx) || (idx >= data.children.size()) || (data.children[idx] != p_child)) {
+		idx = -1;
+
+		// Reget the childen, the list may have changed...
+		children = data.children.ptrw();
+
+		if (p_child->data.pos >= 0 && p_child->data.pos < child_count) {
+			if (children[p_child->data.pos] == p_child) {
+				idx = p_child->data.pos;
+			}
+		}
+
+		if (idx == -1) { //maybe removed while unparenting or something and index was not updated, so just in case the above fails, try this.
+			for (int i = 0; i < child_count; i++) {
+				if (children[i] == p_child) {
+					idx = i;
+					break;
+				}
+			}
+		}
+
+		ERR_FAIL_COND_MSG(idx == -1, vformat("Cannot remove child node '%s' as it is no longer a child of this node.", p_child->get_name()));
```

Comment:
Is this safe in this context? I might be misunderstanding this change, it seems that if we mess this up we can't really be sure that p_child is still pointing to what we think it is?

Reply:
`p_child` is still pointing to the correct child, the issue is that the `idx` can change during the pre-deletion (i.e. the child list can change), which is the mechanism of the bug in the issue.

The `ERR_FAIL` here should not happen unless something *really* bad is going on, like `p_child` is being removed as a child during the pre-delete. This is very likely not possible, but this `ERR_FAIL` is more a fail safe to prevent a crash in this scenario, and indicate where the problem is occurring. As this situation should be pretty rare anyway, there's no cost to adding an `ERR_FAIL` here.

Reply:
Ah there is a potential bug here though, I forgot to reget `get_child` as it could have changed. Will fix. :+1:

Reply:
Just to reiterate, `p_child` will still point to the same child in the same area of memory, as children are non-relocatable - it is only the index that may change.